### ISSUE TITLE
Test jwt_ec fail when gnutls backend is used

### ISSUE THREAD

**Issue by babelouest** [benmcollins/libjwt#76]
I've recently realized that the test `jwt_ec` fails when gnutls backend is enabled (`configure --without-openssl`).
The test output is the following:

```
Running suite(s): LibJWT EC Sign/Verify
14%: Checks: 7, Failures: 6, Errors: 0
jwt_ec.c:89:F:jwt_ec:test_jwt_encode_es256:0: Assertion 'ret == 0' failed: ret == 22, 0 == 0
jwt_ec.c:89:F:jwt_ec:test_jwt_verify_es256:0: Assertion 'ret == 0' failed: ret == 22, 0 == 0
jwt_ec.c:89:F:jwt_ec:test_jwt_encode_es384:0: Assertion 'ret == 0' failed: ret == 22, 0 == 0
jwt_ec.c:89:F:jwt_ec:test_jwt_verify_es384:0: Assertion 'ret == 0' failed: ret == 22, 0 == 0
jwt_ec.c:89:F:jwt_ec:test_jwt_encode_es512:0: Assertion 'ret == 0' failed: ret == 22, 0 == 0
jwt_ec.c:89:F:jwt_ec:test_jwt_verify_es512:0: Assertion 'ret == 0' failed: ret == 22, 0 == 0
jwt_ec.c:194:P:jwt_ec:test_jwt_encode_ec_with_rsa:0: Passed
```

I was able to narrow it down to the [gnutls_pubkey_verify_data2](https://github.com/benmcollins/libjwt/blob/master/libjwt/jwt-gnutls.c#L335) command which fails with the return value -419 (`GNUTLS_E_INCOMPATIBLE_SIG_WITH_KEY`).

**Comment by babelouest** [comment]
Apparently the problen comes from using GnuTLS 3.6 (Debian sid), while it works with GnuTLS 3.5 (Debian Stretch)

**Comment by benmcollins** [comment]
Yeah. I just tried with v3.5.8 from Ubuntu, and it works. I'll try v3.6.x tomorrow.

**Comment by babelouest** [comment]
I'm afraid I'm stuck with GnuTLS 3.6.x for now.
In all the tests I try between GnuTLS 3.5 and GnuTLS 3.6, the ECDSA signing/verifying works with 3.5 but fails with 3.6. It even looks like GnuTLS 3.6 doesn't test ECDSA signing/verifying:
- https://gitlab.com/gnutls/gnutls/blob/master/tests/x509sign-verify-ecdsa.c
- https://gitlab.com/gnutls/gnutls/blob/master/tests/x509sign-verify-common.h#L194

If I force testing ECDSA, the test fails...

I asked for help on the mailing list, waiting for some answers.

**Comment by benmcollins** [comment]
Sounds like you're saying this is actually a GnuTLS bug?

**Comment by babelouest** [comment]
I'm afraid so...
I'll wait for the upstream opinion though.

**Comment by babelouest** [comment]
Hello @benmcollins ,

Quick question about this fix. Do you plan on releasing a new libjwt version soon?
I'm wondering because I'm trying to update the Debian package `libjwt` with both openssl and gnutls backend available but this bug was blocking the multi backend packages purpose.

Now that I know how to fix it, I can either wait for the next release to have the fix in the official source, or patch the debian package to fix the gnutls issue.

**Comment by benmcollins** [comment]
I'll kick out a release this weekend for you!

**Comment by babelouest** [comment]
I also thought that tests could be improved with failure tests. I can make some new tests cases this week-end, if you want it to be ready for the end of the year :)

**Comment by benmcollins** [comment]
Much appreciated. There are a few tests with expected failures in order to get some code coverage on the error paths. Some of the are unreachable without some catastrophic failure (e.g. malloc returning NULL).
